### PR TITLE
Add max-chunks-bytes-per-query limiter

### LINKED ISSUE #3669: Limit series per query in the blocks storage
**Is your feature request related to a problem? Please describe.**
The chunks storage support the `-ingester.max-series-per-query` limit, which allows to configure the maximum number of series for which a query can fetch samples from ingesters. It's not a full support to limit series per query, because the limit is enforced in the ingesters only.

Blocks storage doesn't support it at all.

**Describe the solution you'd like**
Ideally, we should limit series per query in the blocks storage, both when fetching samples from ingesters and long term storage. The store-gateway internally supports a series limit (we're currently not using it), but that would be enforced in a single store-gateway only. We could do it, plus enforced in ingesters and queriers, to have a real limit implementation.

### PR DESCRIPTION
**Pull request by treid314**
Signed-off-by: Tyler Reid <tyler.reid@grafana.com>

**What this PR does**:
This PR adds a new `-querier.max-chunk-bytes-per-query` limit to limit the amount of bytes a query can use for storing chunks for a single query. 

**Which issue(s) this PR fixes**:
Fixes #3669

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/util/limiter/query_limiter.go
Comment (treid314) [inline pkg/util/limiter/query_limiter.go:31]: This limits us to 2GB (2^31 -1 bytes) per query, is it worth making this an unsigned int which is about 4GB (2^32 bytes) per query or a 64 bit number?

File: pkg/distributor/query.go
Comment (treid314) [inline pkg/distributor/query.go:241]: Do we want to divide the size here by the replication factor to handle replicated series on the ingester?

File: pkg/util/validation/limits.go
Comment (pracucci) [inline pkg/util/validation/limits.go:151]: Few things:
1. The CLI flag name is wrong (there's a white space)
2. I would call it `querier.max-fetched-chunk-bytes-per-query` to keep it consistent with `querier.max-fetched-chunks-per-query`. Please rename the YAML field and CHANGELOG accordingly and then run `make doc` to update the auto-generated doc.

Comment (pracucci) [reply pkg/util/validation/limits.go:151]: > The maximum number of chunk bytes for which a query can fetch from each ingesters and blocks storage. This limit is enforced in the querier only when running Cortex with blocks storage. 0 to disable

What do you think if we try to simplify it as:

The maximum aggregated size of all chunks that can be fetched from ingesters and long-term storage per query. This limit is enforced in the querier only when running Cortex with blocks storage. 0 to disable.

_Please get used to add dots (.) at the end of the last sentence._

File: pkg/distributor/query.go
Comment (pracucci) [reply pkg/distributor/query.go:241]: I don't think so. The limit is an hard cap on the chunks loaded in memory, so shouldn't be divided by the replication factor.

File: pkg/distributor/query.go
Comment (pracucci) [inline pkg/distributor/query.go:240]: You can add a function on `TimeSeriesChunk` (in `pkg/ingester/client/custom.go`) to count the total size of all chunks in `TimeSeriesChunk` and then call `queryLimiter.AddChunkBytes()` just once. This will simplify the logic here.

File: pkg/querier/blocks_store_queryable.go
Comment (pracucci) [inline pkg/querier/blocks_store_queryable.go:59]: This error shouldn't be here.

File: pkg/querier/blocks_store_queryable.go
Comment (pracucci) [inline pkg/querier/blocks_store_queryable.go:631]: Same comment here. I would count the total size of all chunks and then call `AddChunkBytes()` once.

File: pkg/util/limiter/query_limiter.go
Comment (pracucci) [inline pkg/util/limiter/query_limiter.go:21]:
```suggestion
	errMaxChunkBytesHit = "The query hit the max aggregated chunks size limit (limit: %d bytes)"
```

File: pkg/util/limiter/query_limiter.go
Comment (pracucci) [inline pkg/util/limiter/query_limiter.go:36]: [nit]

```suggestion
func NewQueryLimiter(maxSeriesPerQuery, maxChunkBytesPerQuery int) *QueryLimiter {
```

File: pkg/distributor/query.go
Comment (pracucci) [inline pkg/distributor/query.go:240]: Can we have a unit test on this too?

**Review by pracucci — COMMENTED** [review]
Good job! I left few comments but overall logic LGTM 👏

File: pkg/util/limiter/query_limiter.go
Comment (pstibrany) [reply pkg/util/limiter/query_limiter.go:31]: On 64-bit systems, `int` is 64-bit, so this is fine. Note that Cortex officially doesn't support 32-bit systems.

Comment (pracucci) [reply pkg/util/limiter/query_limiter.go:31]: I would be explicit like we do everywhere else.

File: CHANGELOG.md
Comment (pracucci) [inline CHANGELOG.md:11]: [nit]

```suggestion
* [FEATURE] Querier: Added new `-querier.max-fetched-chunk-bytes-per-query` flag. When Cortex is running with blocks storage, the max chunk bytes limit is enforced in the querier and limits the size of all aggregated chunks returned from ingesters and blocks storage as bytes for a query. #4216
```

File: pkg/util/limiter/query_limiter.go
Comment (pracucci) [inline pkg/util/limiter/query_limiter.go:21]: You can export it so you can use it in `TestBlocksStoreQuerier_Select` instead of copy-pasting the error string there. Same can be done for `errMaxSeriesHit`.

As already suggested in a previous comment, I would make explicit the unit:

```suggestion
	errMaxChunkBytesHit = "The query hit the aggregated chunks size limit (limit: %d bytes)"
```

File: pkg/distributor/distributor_test.go
Comment (pracucci) [inline pkg/distributor/distributor_test.go:1015]: This comment is incorrect. We're not testing the max series limit here.

File: pkg/distributor/distributor_test.go
Comment (pracucci) [inline pkg/distributor/distributor_test.go:1016]: How was 60 computed? I think it's safer if you do something like this:

1. Call `writeReq := makeWriteRequest(0, 10, 0)` to generate a request with 10 series
2. Calculate the chunks size, multiply by replication factor, add +1: that's the limit
3. Query it back: should succeed
4. Add 1 more series
5. Query it back: should fail

File: pkg/distributor/distributor_test.go
Comment (pracucci) [inline pkg/distributor/distributor_test.go:1027]: This comment is incorrect.

File: pkg/distributor/distributor_test.go
Comment (pracucci) [inline pkg/distributor/distributor_test.go:1043]: This comment is incorrect.

File: pkg/util/limiter/query_limiter.go
Comment (treid314) [reply pkg/util/limiter/query_limiter.go:31]: Should we also pass in an int64 at the config/limit.go level? Or is leaving NewQueryLimiter(int, int) and casting the maxChunkBytes value to an int64 ok?

Comment (pstibrany) [reply pkg/util/limiter/query_limiter.go:31]: > I would be explicit like we do everywhere else.

I don't think we're explicit "everywhere else". I think it would make sense to use `int` here simply because we cannot fit more than `max of int` into memory anyway (applies for both 32-bit and 64-bit platforms).

Comment (pstibrany) [reply pkg/util/limiter/query_limiter.go:31]: To your question Tyler, if you go with `int64` route, you will need to "extend" that everywhere to avoid losing precision somewhere (ie. in `NewQueryLimiter` too)

Comment (pracucci) [reply pkg/util/limiter/query_limiter.go:31]: Ok. Let's not block on this and keep `int`.

File: pkg/distributor/distributor_test.go
Comment (pracucci) [inline pkg/distributor/distributor_test.go:955]: You just made me realise that `ctx` is global, but shouldn't. Instead of using `t.Cleanup()` please remove the `ctx` global variable and define it in each test that needs it.

File: pkg/distributor/distributor_test.go
Comment (pracucci) [inline pkg/distributor/distributor_test.go:1043]: Remove this too (see previous message).

File: pkg/distributor/distributor_test.go
Comment (pracucci) [inline pkg/distributor/distributor_test.go:1008]: You can remove this, and just add `seriesToAdd-1` below, after the 1st series has been pushed.

File: pkg/distributor/distributor_test.go
Comment (pracucci) [inline pkg/distributor/distributor_test.go:1063]: You have already pushed `another_series` with the same timestamp above. In a real world scenario, this push will fail because of duplicated timestamp (not allowed in ingesters). However, here we're using a mocked ingester and so it works (but it took me a good 5 minutes to understand why). So, I would call this series different to not trick the next reader too.

**Review by pracucci — COMMENTED** [review]
Thanks a lot Tyler to address my feedback! I think the PR logic is good to go. I just have few last comments on tests that I would be glad to see addressed before merging. Thanks! 🚀

File: pkg/util/limiter/query_limiter.go
Comment (pstibrany) [inline pkg/util/limiter/query_limiter.go:28]: nit: When embedding atomic into other structs, and these structs are only accessed via pointers, atomic itself doesn't need to be a pointer.

In other words, `chunkBytesCount atomic.Int64` would work fine here, and then there is no need to call `atomic.NewInt64(0)` in `NewQueryLimiter`.

File: pkg/util/validation/limits.go
Comment (pstibrany) [inline pkg/util/validation/limits.go:151]:
```suggestion
	f.IntVar(&l.MaxFetchedChunkBytesPerQuery, "querier.max-fetched-chunk-bytes-per-query", 0, "The maximum size of all chunks in bytes that a query can fetch from each ingester and storage. This limit is enforced in the querier and ruler only when running Cortex with blocks storage. 0 to disable.")
```

File: CHANGELOG.md
Comment (pstibrany) [inline CHANGELOG.md:11]:
```suggestion
* [FEATURE] Querier/Ruler: Added new `-querier.max-fetched-chunk-bytes-per-query` flag. When Cortex is running with blocks storage, the max chunk bytes limit is enforced in the querier and ruler and limits the size of all aggregated chunks returned from ingesters and storage for a query. #4216
```

File: pkg/util/limiter/query_limiter.go
Comment (pstibrany) [inline pkg/util/limiter/query_limiter.go:95]: Nit: Same comment as in `AddSeries` -- no need to return `validation.LimitError` from here. Simple  `return fmt.Sprintf(ErrMaxChunkBytesHit, ql.maxChunkBytesPerQuery)` would remove dependency on `validation` package. Calling code (querier package) can add this wrapping when needed.

**Review by pstibrany — APPROVED** [review]
Nice work! I've left few nit comments (mention ruler in the changelog/help, remove duplicite mentions of blocks storage).

File: pkg/util/limiter/query_limiter.go
Comment (pracucci) [reply pkg/util/limiter/query_limiter.go:95]: I jump in on this. It's in the TODO list, but I suggested to do it in a follow up PR to keep changes easier to review.

File: pkg/distributor/ha_tracker_test.go
Comment (pracucci) [inline pkg/distributor/ha_tracker_test.go:664]: [nit] `userID`.

**Review by pracucci — APPROVED** [review]
Thanks for addressing my feedback! One final nit and we go! 🚀 🌔

**Pull request merged**
